### PR TITLE
Fix bug in sizeWithFont

### PR DESCRIPTION
**Pull request by aballway**
NSString+UIKitAdditions sizeWithFont:constrainedToSize: should return the height that the text requires rather than the height of the text that can actually fit in the region that CTFramesetterSuggestFrameSizeWithConstraints returns.

### CODE REVIEW

File: Frameworks/UIKit/NSString+UIKitAdditions.mm [hunk 1]
```diff
@@ -259,9 +259,8 @@ - (CGSize)_sizeWithAttributes:(NSDictionary<NSString*, id>*)attributes constrain
         size.width = std::numeric_limits<CGFloat>::max();
     }
 
-    if (size.height == 0.0) {
```
Comment (yiyang-msft) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: >if (size.height == 0.0) { [](start = 4, length = 25)

no wonder with this, it works. :)

**Comment by yiyang-msft** [comment]
:shipit:

File: Frameworks/UIKit/NSString+UIKitAdditions.mm [hunk 2]
```diff
@@ -259,9 +259,8 @@ - (CGSize)_sizeWithAttributes:(NSDictionary<NSString*, id>*)attributes constrain
...
-        size.height = std::numeric_limits<CGFloat>::max();
-    }
+    // Return the height that the text actually requires rather than the constrained size
+    size.height = std::numeric_limits<CGFloat>::max();
```
Comment (ms-jihua) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: so wait the expectation of the constrainedToSize: contract is that the height passed in doesn't matter at all? that seems... wrong...

Comment (jaredhms) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: that does seem a bit odd

---
In reply to: [99448184](https://github.com/Microsoft/WinObjC/pull/1917#discussion_r99448184) [](ancestors = 99448184)

Comment (aballway) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: @yiyang-msft and I reproduced this on the reference platform, but it's likely that this only applies to a case where we are only drawing one line but unable to fit it.  I'm going to be doing more testing to see where height is respected and where it is not, and if this also applies to CTFramesetterSuggestFrameSizeWithConstraints.

File: tests/unittests/UIKit/NSString+UIKitAdditionsTests.mm [hunk 3]
```diff
@@ -17,9 +17,21 @@
 #import <TestFramework.h>
 #import <UIKit/UIKit.h>
 
+static constexpr double c_errorDelta = .0001;
```
Comment (rajsesh-msft) [inline tests/unittests/UIKit/NSString+UIKitAdditionsTests.mm]: nit: use numEquals

**Comment by jaredhms** [comment]
:shipit:

File: Frameworks/UIKit/NSString+UIKitAdditions.mm (hunk 2 above)
Comment (yiyang-msft) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: @alballw@microsoft.com , make sense.

---
In reply to: [99671953](https://github.com/Microsoft/WinObjC/pull/1917#discussion_r99671953) [](ancestors = 99671953)

Comment (yiyang-msft) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: also,please verified with MM. check with the introduction label on first screen.

---
In reply to: [99942712](https://github.com/Microsoft/WinObjC/pull/1917#discussion_r99942712) [](ancestors = 99942712,99671953)

Comment (aballway) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: With this change the introduction label in Islandwood house is being scaled properly

---
In reply to: [99943020](https://github.com/Microsoft/WinObjC/pull/1917#discussion_r99943020) [](ancestors = 99943020,99942712,99671953)

File: Frameworks/UIKit/NSString+UIKitAdditions.mm [hunk 4]
```diff
@@ -263,7 +263,14 @@ - (CGSize)_sizeWithAttributes:(NSDictionary<NSString*, id>*)attributes constrain
         size.height = std::numeric_limits<CGFloat>::max();
     }
 
-    return CTFramesetterSuggestFrameSizeWithConstraints(framesetter.get(), CFRangeMake(0, self.length), nullptr, size, nullptr);
+    CGSize ret = CTFramesetterSuggestFrameSizeWithConstraints(framesetter.get(), CFRangeMake(0, self.length), nullptr, size, nullptr);
+
+    // If the constrained height is less than a line height sizeWithFont will increase the returned size to fit at least one line
+    UIFont* font = attributes[NSFontAttributeName];
```
Comment (DHowett-MSFT) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: What if there's no font?

Comment (jaredhms) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: i figured you'd just end up with a 0 size, so you'd get ret's height.

Comment (DHowett-MSFT) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: If the compiler is working as intended, yes :) We had issues with returning floats from nil receivers before.

Comment (aballway) [inline Frameworks/UIKit/NSString+UIKitAdditions.mm]: Looks like reference platform just returns CGSizeZero, will update to reflect that

**Pull request merged**